src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -650,6 +654,113 @@ def process_declarations(self, parent, nodes, arg_list):
+           isinstance(intrinsics[0].items[0],
+                      Fortran2003.Intrinsic_Name) and \
+                      str(intrinsics[0].items[0]).lower() == "kind":
+            # We have kind=KIND(X) where X may be of any intrinsic type. It
+            # may be a scalar or an array. items[1] is an
+            # Actual_Arg_Spec_List with the first entry being the argument.
+            kind_arg = intrinsics[0].items[1].items[0]
+
+            # We currently only support literals as arguments to KIND
+            if isinstance(kind_arg, (Fortran2003.Int_Literal_Constant,
+                                     Fortran2003.Real_Literal_Constant)):
+
+                if len(kind_arg.items) > 1 and kind_arg.items[1]:
+                    # The literal has an explicit kind specifier (e.g. 1.0_wp)
+                    # so return a Symbol representing it (e.g.'wp').
+                    return Fparser2Reader._kind_symbol_from_name(
+                        str(kind_arg.items[1]), symbol_table)
+
+                if isinstance(kind_arg, Fortran2003.Real_Literal_Constant):
+                    import re
+                    if re.search("d[0-9]", str(kind_arg).lower()):
+                        return Symbol.Precision.DOUBLE
+                    return Symbol.Precision.SINGLE
+
+                if isinstance(kind_arg, Fortran2003.Int_Literal_Constant):
+                    # An integer with no explict kind specifier must be of
+                    # default precision
+                    return Symbol.Precision.SINGLE
+
+            raise NotImplementedError(
+                "Only real and integer literals are supported "
+                "as arguments to the KIND intrinsic but found '{0}' in: "
+                "{1}".format(type(kind_arg).__name__, str(kind_selector)))
+
+        # We have kind=kind-param
+        kind_names = walk_ast(kind_selector.items, [Fortran2003.Name])
+        if not kind_names:
+            raise NotImplementedError(
+                "Failed to find valid Name in Fortran Kind "
+                "Selector: '{0}'".format(str(kind_selector)))
+        return Fparser2Reader._kind_symbol_from_name(str(kind_names[0]),
+                                                     symbol_table)
+
+    @staticmethod
+    def _kind_symbol_from_name(name, symbol_table):
+        '''
+        Utility method that returns a Symbol representing the named KIND
+        parameter. If the supplied Symbol Table does not contain an appropriate
+        entry then one is created. If it does contain a matching entry then
+        its datatype must be 'integer' or 'deferred'. If the latter then the
+        fact that we now know that this Symbol represents a KIND parameter
+        means that we can change the datatype to be 'integer'.
+
+        :param str name: the name of the variable holding the KIND value.
+        :param symbol_table: the Symbol Table associated with the code being\
+                             processed.
+
```

Comment:
Type of symbol_table should be specified.

Reply:
Oh yes. I'd forgotten to document the return value too.